chain/transaction.go

```diff
@@ -361,6 +372,20 @@ func (t *Transaction) Execute(
 
 		actionOutputs = append(actionOutputs, encodedOutput)
 	}
+
+	// We refund here
+	refund, err := scope.Refund(r)
+	if err != nil {
+		return nil, ErrRefundDimensions
+	}
+	refundFee, err := feeManager.Fee(refund)
+	if err != nil {
+		return nil, ErrFailedToComputeRefund
+	}
+	if err := bh.AddBalance(ctx, t.Auth.Sponsor(), ts, refundFee); err != nil {
+		return nil, ErrRefundFailed
+	}
```

Comment:
Don't we have all the information needed to handle the refund immediately after fetching, so we can differentiate between tiered storage accessed? Do we need to wait until after execution for some reason?

Reply:
Keeping the refund logic inside of `transaction.Execute()` would be consistent with how we're currently manipulating with the sponsor's balance. It's true that once we've fetched all values from the fetcher (in both the block builder and the validator) that we can immediately compute/apply the refund. However, this would require us to make a call to `BalanceHandler` outside of `transaction.Execute()`, which could be considered an antipattern as the sponsor's balance is manipulated only in `transaction.Execute()`.

As where the refund logic is placed inside `transaction.Execute()`, we could apply the refund after deducting. A property we gain from the current implementation is that refunds are applied only to successful transactions, but I wouldn't mind removing it.